### PR TITLE
Comments fully implemented

### LINKED ISSUE #65: Create comment section attached to each main content view
We need a comment section for our main content view. Use a rich text editor such as slate. We will need to create a new collection, which will have a schema sort of like this:

{
Id: string 
Type: enum (FI, SF, PPR, UPR)
comment: string
}
The comment will store a blob text thats generated by a rich text editor such as slate. For any given main content page, we should show comments in reverse chronological order (most recent at front)

### PR DESCRIPTION
**Pull request by willi-li-am**
Used slate-react to input and show comments
Added new route called '/comments'
 - post, creates new comment (no form validation on backend)
 - get, gets all comments + replies with ticket id
   - Aggregates replies in an array in the response
 - No edits and deletion yet

Every comment is stored in comments (replies too)
 - Reference code is either _id of parent comment or ticket_id

No error toasts yet

Had a few issues with getting the comments to update with slate and emptying the input, so whenever function getComments in frontend is called, it forcefully rerenders the whole comment section

Also had issues with placeholder text, since it wouldn't follow the padding I gave to the text area
 - decided to remove placeholder altogether

resolves #65

### CODE REVIEW

File: backend/models/comment.model.js [hunk 1]
```diff
@@ -1,11 +1,15 @@
 const mongoose = require('mongoose')
 const { Schema } = mongoose
 
+const Mixed = Schema.Types.Mixed
+
+//sort by new -> old for parent comments, old -> new for replies
```
Comment (victorzheng02) [inline backend/models/comment.model.js]: I get what this comment means, but why in this specific file?

File: backend/routes/comments.routes.js [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+const {
+    createCommentController,
+    getAllCommentsController,
+} = require('../controller/comments.controller')
+
+router.route('/:id').get(authenticateUser, getAllCommentsController)
+
+router.route('/').post(authenticateUser, createCommentController)
+
+//delete, edit and others can come later
```
Comment (victorzheng02) [inline backend/routes/comments.routes.js]: prefix with TODO: ideally so its easier to find

File: backend/service/comments.service.js [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+const getAllComments = async (code) => {
+    //add reply aggregation, sortby
+    const res = await Comment.aggregate([
+        {
+            $match: {
+                reference_code: code,
+            },
+        },
+        {
+            $lookup: {
```
Comment (victorzheng02) [inline backend/service/comments.service.js:18]: we love aggregation pipelines 😍😍😍

File: frontend/public/index.html [hunk 4]
```diff
@@ -1,4 +1,4 @@
-<!DOCTYPE html>
+<!doctype html>
```
Comment (victorzheng02) [inline frontend/public/index.html]: revert this change

File: frontend/src/components/CommentInput.js [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+    toggleMark,
+} from './SlateComponents'
+
+const HOTKEYS = {
+    'mod+b': 'bold',
+    'mod+i': 'italic',
+    'mod+u': 'underline',
+}
+
+const cleanInput = (val) => {
```
Comment (victorzheng02) [inline frontend/src/components/CommentInput.js:26]: mind leaving a comment so it's clear whats happening here

File: frontend/src/components/CommentInput.js [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+    const renderLeaf = useCallback((props) => <Leaf {...props} />, [])
+    const editor = useMemo(() => withHistory(withReact(createEditor())), [])
+    const [loading, setLoading] = useState(false)
+    const auth = useAuth()
+    const [val, setVal] = useState(editor.children)
+    const handleSubmit = (ref, ticket) => {
+        setLoading(true)
+        const comment = cleanInput(val)
+        if (comment.length === 0) {
+            console.log('EMPTY')
```
Comment (victorzheng02) [inline frontend/src/components/CommentInput.js]: remove console log (theres one more in the pr later as well)

File: frontend/src/components/CommentView.js [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+    Popover,
+    PopoverTrigger,
+    PopoverContent,
+    HStack,
+} from '@chakra-ui/react'
+
+import { Element, Leaf } from './SlateComponents'
+import CommentInput from './CommentInput'
+
+const timeAgo = (dateString) => {
```
Comment (victorzheng02) [inline frontend/src/components/CommentView.js]: we should probably move this into a helpers file, as this sort of calculation/formatting might be used elsewhere and shouldn't be specific to this file

**Review by victorzheng02 — APPROVED** [review]
this is fire, just had a few small comments

File: frontend/src/components/CommentInput.js [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+        }
+    }
+    for (let i = l; i <= r; i++) {
+        res.push(val[i])
+    }
+    return res
+}
+
+const invalidInput = (val) => {
+    if (val.length === 0) {
```
Comment (ansonjwhe) [inline frontend/src/components/CommentInput.js]: seems like a redundant check. if `val.length` is 0 then the for loop below shouldn't run

File: frontend/src/components/CommentSection.js [hunk 9]
```diff
@@ -1,114 +1,89 @@
...
+        <Box mt="50px" mb="50px">
+            <Box
+                fontWeight="bold"
+                fontSize="30px"
+                textAlign="left"
+                marginLeft="5px"
+            >
+                Comments
+            </Box>
+            <div key={refreshKey}>
```
Comment (ansonjwhe) [inline frontend/src/components/CommentSection.js:48]: what is the purpose of this `refreshKey`?

File: frontend/src/components/CommentView.js [hunk 10]
```diff
@@ -0,0 +1,275 @@
...
+
+    if (seconds < 60) {
+        return 'Just now'
+    } else if (minutes < 60) {
+        return `${minutes} minute${minutes > 1 ? 's' : ''} ago`
+    } else if (hours < 24) {
+        return `${hours} hour${hours > 1 ? 's' : ''} ago`
+    } else if (days < 7) {
+        return `${days} day${days > 1 ? 's' : ''} ago`
+    } else if (weeks < 4.35) {
```
Comment (ansonjwhe) [inline frontend/src/components/CommentView.js]: where does this 4.35 come from? is this supposed to represent 30 days in a month?

File: frontend/src/components/CommentSection.js (hunk 9 above)
Comment (willi-li-am) [reply frontend/src/components/CommentSection.js:48]: Slate editor doesn't rerender the actual comment values once getComments() is called again when mapping through comments

{comments.map((content, index) => {
    return (
        <CommentView
            key={
                'comment-' +
                (comments.length - index).toString() +
                ticket
            }
            comment={content}
            allUsers={allUsers}
            getComments={getComments}
        />
    )
})}

Comment (willi-li-am) [reply frontend/src/components/CommentSection.js:48]: So I opted to force rerender to change the actual values
